components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -580,6 +592,224 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
+            return scopes;
+        }
+
+        // Get the application requested scopes for the federated tokens.
+        String requestedScopesForTokenSharing = getRequestedScopesForTokenSharing(context);
+
+        // Validating the application requested scopes by the authenticator allowed scopes for federated token sharing.
+        String validScopesForTokenSharing = validateScopeForTokenSharing(
+                authenticatorProperties.get(OIDCAuthenticatorConstants.FEDERATED_TOKEN_ALLOWED_SCOPE),
+                requestedScopesForTokenSharing);
+
+        if (StringUtils.isNotBlank(validScopesForTokenSharing)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Valid scopes matched for federated token sharing: " + validScopesForTokenSharing);
+            }
+            scopes = (scopes + SPACE + validScopesForTokenSharing).trim();
+        }
+        return scopes;
+    }
+
+    /**
+     * This method returns the scopes requested by the application for the federated tokens.
+     *
+     * @param context The authentication context.
+     * @return The scopes requested by the application for token sharing.
+     */
+    private String getRequestedScopesForTokenSharing(AuthenticationContext context) {
+        // The first priority is given to the parameters setup at the adaptive script. Then the query parameters.
+        String requestedScopesViaAdaptiveScript =
+                getAdaptiveScriptValues(context, OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE);
+        // Checks if there exists scopes requested via adaptive script.
+        if (StringUtils.isNotBlank(requestedScopesViaAdaptiveScript)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Adaptive script parameter found for " + OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE);
+            }
+            return requestedScopesViaAdaptiveScript;
+        } else {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("No adaptive script parameter found for "
+                        + OIDCAuthenticatorConstants.FEDERATED_TOKEN_SCOPE + " hence checking the query parameters.");
+            }
+            // Checks the scopes requested via authorize request query parameters.
+            return getRequestedScopedViaQueryParams(context);
+        }
+    }
+
+    /**
+     * This method returns the adaptive script federated authenticator param value for a given parameter name.
+     *
+     * @param context                The authentication context with federated authenticator params.
+     * @param authenticatorParamName The federated authenticator parameter name.
+     * @return The adaptive script federated authenticator param value for the given parameter name.
+     */
+    private String getAdaptiveScriptValues(AuthenticationContext context, String authenticatorParamName) {
+
+        Map<String, String> runtimeParams = this.getRuntimeParams(context);
+        String requestedScopesViaAdaptiveScript = null;
```

Comment:
If this method is not considered a general utility function to retrieve the params in the adaptive script, shall we refer to the parameter constant itself without passing? 
Otherwise, let's rename this variable for a general value without bounding to the `requestedScopes`

Reply:
Fixed 473c2f0d7c5775dee2cb9ced70449de004c3184f